Chaining side effects (callbacks and stopping)

raboof opened the pull request:
Cleans up the API in a couple of places, but still needs type parameters
in spots where I wouldn't want to need them...

raboof committed:
Chaining side effects (callbacks and stopping)

Cleans up the API in a couple of places, but still needs type parameters
in spots where I wouldn't want to need them...

raboof commented on akka-typed-tests/src/test/scala/akka/typed/persistence/scaladsl/PersistentActorCompileOnlyTest.scala:
```diff
@@ -378,4 +378,29 @@ object PersistentActorCompileOnlyTest {
+          case Enough ⇒ Persist(Done)
+            .andThen((_: State) ⇒ println("yay"))
+            // Ugh it doesn't make much sense to have to include the type parameters here...
+            .andThen(Stop[Event, State]())
```
I couldn't figure out how to make the type inference propagate nicely yet...

raboof replied:
cop-out approach is of course to introduce `andThenStop` (`def andThenStop: PersistentEffect[Event, State] = andThen(Stop[Event, State]())`), but I wonder if there's a better way

patriknw replied:
does `Nothing` work? then we could have a `val Stop`

could also try `def Stop[Event, State]`

patriknw replied:
by the way, `Event` shouldn't be need in `PersistentEffect`, only `State`

raboof replied:
I think we do need `Event`, as we want to make sure `Persist` is only called with `Event`s.

It's kind of interesting that no effect seems to need both `Event` and `State` though, only one or the other.

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/internal/PersistentActorImpl.scala:
```diff
@@ -117,5 +103,27 @@ import akka.typed.internal.adapter.ActorRefAdapter
+      // the invalid event, in case such validation is implemented in the event handler.
+      state = applyEvent(state, event)
+      persist(event) { _ ⇒
+        callback.foreach(_.apply(()))
```
this `callback` isn't used now, I think

raboof replied:
It's filled in the recursive call when handling the `CompositeEffect`

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -20,20 +20,29 @@ object PersistentActor {
-    def andThen(callback: State ⇒ Unit) = copy(callbacks = callback :: callbacks)
+  case class CompositeEffect[Event, State](mainEffect: PersistentEffect[Event, State], sideEffects: List[ChainableEffect[_, State]]) extends PersistentEffect[Event, State] {
+    override def andThen(sideEffect: ChainableEffect[_, State]): PersistentEffect[Event, State] =
+      copy(sideEffects = sideEffects :+ sideEffect)
```
could we prepend (since we use List, and that is good since we expect few elements), and reverse before executing?

patriknw replied:
could something like this be simpler?
```
sealed abstract class PersistentEffect[Event, State]() {
  def andThen(effect: PersistentEffect[Event, State]): PersistentEffect[Event, State] =
      CompositeEffect(effect :: this: Nil)
}

@InternalApi private[akka] final case class CompositeEffect[Event, State](effects: List[PersistentEffect[Event, State]])
    extends PersistentEffect[Event, State] {
    override def andThen(effect: PersistentEffect[Event, State]): PersistentEffect[Event, State] =
      CompositeEffect(effect :: effects)
  }
```

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -20,20 +20,29 @@ object PersistentActor {
+  case class PersistNothing[Event, State]() extends PersistentEffect[Event, State]
+  case class Persist[Event, State](event: Event) extends PersistentEffect[Event, State]
+
+  trait ChainableEffect[Event, State] {
```
do we need to make a difference?
Persist(evt1).andThen(Persist(evt2)) is not impossible, or wrong

raboof replied:
My thinking here was that it might help to separate the `Persist` (which depends on `[Event]` but is independent of `[State]`) from the `SideEffect` (which depends on `[State]` but is independent of `[Event]`) might help keeping the types inferrable. 

Also, do we want to allow `Persist(evt1).andThen(SideEffect(effect1)).andThen(Persist(evt2))`? That might make the implementation a bit harder, and I'm not sure it's something we'd want to promote.

patriknw replied:
I don't see any bad consequences of allowing such chaining. Isn't it like a depth first traversal of a tree structure? We don't expect any deep structures so don't have to worry about recursion depth.

raboof replied:
It does mean making 2 database calls (in the mean time 'holding up' processing new commands), but indeed it might not be a problem

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/internal/PersistentActorImpl.scala:
```diff
@@ -117,5 +103,27 @@ import akka.typed.internal.adapter.ActorRefAdapter
+      // apply the event before persist so that validation exception is handled before persisting
+      // the invalid event, in case such validation is implemented in the event handler.
+      state = applyEvent(state, event)
+      persist(event) { _ ⇒
```
one could continue applying remaining effects here

raboof committed:
Allow multiple side effects in one 'andThen'

Making type inference actually work

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -20,17 +20,17 @@ object PersistentActor {
-  case class CompositeEffect[Event, State](mainEffect: PersistentEffect[Event, State], sideEffects: List[ChainableEffect[_, State]]) extends PersistentEffect[Event, State] {
-    override def andThen(sideEffect: ChainableEffect[_, State]): PersistentEffect[Event, State] =
-      copy(sideEffects = sideEffects :+ sideEffect)
+  case class CompositeEffect[Event, State](mainEffect: PersistentEffect[Event, State], effects: Seq[ChainableEffect[_, State]]) extends PersistentEffect[Event, State] {
```
internal api

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -20,20 +20,29 @@ object PersistentActor {
 
-  case class Persist[Event, State](event: Event, callbacks: List[State ⇒ Unit] = Nil) extends PersistentEffect[Event, State] {
-    def andThen(callback: State ⇒ Unit) = copy(callbacks = callback :: callbacks)
+  case class CompositeEffect[Event, State](mainEffect: PersistentEffect[Event, State], effects: Seq[ChainableEffect[_, State]]) extends PersistentEffect[Event, State] {
```
internal api?

raboof committed:
Allow applying multiple side effects

In a way that allows 'inspecting' the effects if you want to, but
doesn't require you to repeat generic parameters when you don't.

raboof commented:
I've kept the distinction between actually persisting and `ChainableEffect`s because, ultimately, creating `andThen(effect: PersistentEffect[Event, State])` inside `PersistentEffect[+Event, State]` is not allowed due to variance. If anyone can find a way to avoid/simplify that (without introducing new problems ofc ;) ) I'd be interested :)

patriknw reviewed (approved):
LGTM, I'll merge and do the final touch in my PR
